crates/ruff_minimizer/src/main.rs

```diff
@@ -0,0 +1,438 @@
+//! Minimize a failing python file.
+//!
+//! See `--help` (or the [Args] docs) for usage.
+//!
+//! ## Example
+//!
+//! Minimize a file with unstable formatting
+//! ```shell
+//! cargo run --bin ruff-minimizer -- target/cpython/Lib/test/inspect_fodder.py target/minirepo/a.py "Unstable formatting" "target/debug/ruff_dev format-dev --stability-check target/minirepo"
+//! ```
+//! This could emit
+//! ```python
+//! class WhichComments:    # before f        return 1        # end f    # after f    # before asyncf - line 108
+//!     async def asyncf(self):        return 2
+//!         # end asyncf    # end of WhichComments - line 114# a closing parenthesis with the opening paren being in another line
+//! (
+//! );
+//! ```
+//!
+//! Minimize a file with a syntax error
+//! ```shell
+//! cargo run --bin ruff_minimizer -- target/checkouts/jhnnsrs:mikro-napari/mikro_napari/models/representation.py target/minirepo/code.py "invalid syntax" "target/debug/ruff_dev format-dev --stability-check target/minirepo"
+//! ```
+//! This could emit
+//! ```python
+//! class RepresentationQtModel():
+//!             data[:,] = rep.data
+//! ```
+
+#![allow(clippy::print_stdout, clippy::print_stderr)]
+
+use anyhow::{Context, Result};
+use clap::Parser;
+use fs_err as fs;
+use regex::Regex;
+use ruff_python_ast::statement_visitor::{walk_body, walk_stmt, StatementVisitor};
+use ruff_python_ast::visitor::{walk_expr, Visitor};
+use rustpython_ast::text_size::TextRange;
+use rustpython_ast::{Expr, Ranged, Stmt, Suite};
+use std::path::{Path, PathBuf};
+use std::process::{Command, ExitCode};
+use std::str;
+use std::time::Instant;
+use tracing::debug;
+
+const STRATEGIES: &[&dyn Strategy] = &[
+    (&StrategyRemoveModuleMember),
+    (&StrategyRemoveStatement),
+    (&StrategyRemoveExpression),
+    (&StrategyReplaceStatementWithPass),
+    (&StrategyRemoveLine),
+    (&StrategyRemoveNewline),
+];
+
+/// Each strategy is a way of producing possible minimizations
+trait Strategy {
+    fn name(&self) -> &'static str;
+
+    fn candidates<'a>(
+        &self,
+        input: &'a str,
+        ast: &'a Suite,
+    ) -> Result<Box<dyn ExactSizeStringIter + 'a>>;
+}
```

Comment:
Nit: This could be implemented as an enum, considering that it is a fixed set of strategies. But I can see how this allows you to split the code nicer.

Reply:
i've had been thinking about an enum too, with traits it's marginally easier to attach methods